core/xml.js

```diff
@@ -462,9 +407,178 @@ Blockly.Xml.domToWorkspace = function(xml, workspace) {
+ * handling or to present more meaningful information to users. The default
+ * behavior is to log the message to the browser console.
+ * @param {!string} message The warning message describing what went wrong.
+ * @public
+ */
+Blockly.Xml.onWarning = function(message) {
+  console.warn(message);
+};
+
+/**
+ * A mapping of XML tags to parser functions.
+ * @type {Object.<string,
+ *     function(Blockly.Workspace, Element, Blockly.Xml.ParserState)>}
+ */
+Blockly.Xml.tagMap_ = {};
+
+/**
+ * A list of operations to serialize a workspace to XML.
+ * @type {Array.<function(Element, Blockly.Workspace, boolean)>}
+ * @private
+ */
+Blockly.Xml.serializers_ = [];
+
+/**
+ * The function to handle unrecognized tags when parsing an XML document.
+ * @param {!Blockly.Workspace} workspace the Blockly workspace
+ * @param {!Element} xmlChild the XML element being parsed
+ * @param {!Blockly.Xml.ParserState} _state parse state information (not used
+ *     in this implementation, but may be used by projects that provide a
+ *     different behavior.
+ * @public
+ */
+Blockly.Xml.unrecognizedTagType = function(workspace, xmlChild, _state) {
+  if (xmlChild.nodeName.toLowerCase() != '#text' &&
+      xmlChild.nodeName.toLowerCase() != '#comment') {
+    Blockly.Xml.onError('Unrecognized tag ' + xmlChild.nodeName.toLowerCase() +
+      ' in XML');
+  }
+};
+
+/**
+ * Shared state used by different tag parsers.
+ * @param {!Blockly.Workspace} workspace the target workspace for parsing
+ * @constructor
+ */
+Blockly.Xml.ParserState = function(workspace) {
+  this.newBlockIds = [];
+  // Not used in LTR.
+  this.width = workspace.RTL ? workspace.getWidth() : undefined;
+  // A list of block IDs added by this parse.
+  this.variablesFirst = true;
+};
+
+/**
+ * Register a handler to process tags at the top-level of the blocks XML DOM.
+ * @param {string} tagName The XML tag name for which to handle the custom
+ *     handler code.
+ * @param {!function(!Blockly.Workspace,!Element,!Blockly.Xml.ParserState)} load
+ *     A function to handle processing the XML Element into the workspace.
+ *     Shared state information is passed as the third argument.
+ * @param {!function(!Element, !Blockly.Workspace, boolean=)} save A
+ *     function to serialize a workspace into the given tag.
+ * @param {boolean=} opt_force Force registering the handler even if it
+ *     will overwrite an existing handler
+ */
+Blockly.Xml.registerTopLevelTag = function(tagName, load, save, opt_force) {
+  if (tagName in Blockly.Xml.tagMap_ && !opt_force) {
+    throw Error('There is a parser already registered for ' + tagName);
+  }
+  Blockly.Xml.tagMap_[tagName] = load;
+  Blockly.Xml.serializers_.push(save);
+};
+
+Blockly.Xml.registerTopLevelTag('variables',
+    function(workspace, xmlChild, state) {
+      if (state.variablesFirst) {
+        Blockly.Xml.domToVariables(xmlChild, workspace);
+      } else {
+        Blockly.Xml.onError('\'variables\' tag must exist once before ' +
```

Comment:
`onError` throws, so if you negate the if condition then you can drop the else, resulting in smaller, simpler code.